csrc/mmdeploy/net/ipu/ipu_net.cpp

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+
+#include "ipu_net.h"
+
+#include <sys/time.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "mmdeploy/core/logger.h"
+#include "mmdeploy/core/model.h"
+#include "mmdeploy/core/utils/formatter.h"
+
+namespace mmdeploy::framework {
+
+mmdeploy::DataType IPUNet::ipu_type_convert(const popef::DataType& ipu_type) {
+  mmdeploy::DataType mtype;
+  if (ipu_type == popef::DataType::F32) {
+    mtype = mmdeploy::DataType::kFLOAT;
+  } else if (ipu_type == popef::DataType::F16) {
+    mtype = mmdeploy::DataType::kHALF;
+  } else if (ipu_type == popef::DataType::F8) {
+    mtype = mmdeploy::DataType::kINT8;
+  } else {
+    throw std::invalid_argument(
+        "invalid data type for IPU backend, current legit is : fp32, fp16, fp8");
+  }
+  return mtype;
+}
+
+void IPUNet::copy_output(const model_runtime::TensorMemory& from, Tensor& to) {
+  if (from.data_size_bytes != to.byte_size()) {
+    MMDEPLOY_ERROR("output tensor size not match from size {} to size {}", from.data_size_bytes,
+                   to.byte_size());
+    return;
+  }
+  int size = from.data_size_bytes;
+
+  MMDEPLOY_DEBUG("copy output total byte_size {}", size);
+
+  char* from_ptr = static_cast<char*>(from.data.get());
+
+  char* pto = to.data<char>();
+  for (int i = 0; i < size; i++) {
+    pto[i] = *(from_ptr + i);
+  }
+}
+
+void IPUNet::copy_input(const Tensor& from, model_runtime::TensorMemory& to) {
+  if (from.byte_size() != to.data_size_bytes) {
+    MMDEPLOY_ERROR("input tensor size not match  from size {} to size {} ", from.byte_size(),
+                   to.data_size_bytes);
+    return;
+  }
+  int size = from.byte_size();
+
+  MMDEPLOY_DEBUG("copy input total byte_size {}", size);
+
+  char* to_ptr = static_cast<char*>(to.data.get());
+
+  const char* pfrom = from.data<char>();
+  for (int i = 0; i < size; i++) {
+    *(to_ptr + i) = pfrom[i];
+  }
+}
+
+IPUNet::~IPUNet() {}
+
+Result<void> IPUNet::Init(const Value& args) {
+  std::string popef_path = args["popef_path"].get<std::string>();
```

Comment:
To load model from sdk, you could refer to https://github.com/gqingraphcore/mmdeploy-ipu/blob/ipu/csrc/mmdeploy/net/trt/trt_net.cpp#L113-L115

Reply:
the C++ interface for IPU can only init from a "converted" model file ("popef") as the converter "poprt" is only implemented in python, so the user have to manually convert an onnx to popef first before using the C++ interface